src/niscope/system_tests/test_system_niscope.py

```diff
@@ -381,6 +382,20 @@ def test_configure_trigger_window(self, multi_instrument_session):
         assert trigger_source == multi_instrument_session.trigger_source
         assert niscope.TriggerWindowMode.ENTERING == multi_instrument_session.trigger_window_mode
 
+    # Multi-Threading tests
+    def test_multi_threading(self, multi_instrument_session):
```

Comment:
I'm having difficulty understanding these tests.
You call commit in two different threads on the same session. But you don't do it at the same time, do you?

I think a reasonable test for the locking would be to:

thread 1:
* lock the session
* get an attribute
* wait 0.5 seconds
* unlock the session

thread 2:
* lock the session
* get an attribute
* wait 0.5 seconds
* unlock the session

Launch both threads as close to each other as you can, confirm total time is at least 1 second and maybe that one attriubte you could get right away but the other one took about 0.5 seconds.

Reply:
FYI, timing based tests are often flaky, especially in PR/CI pipelines. 

Some other approaches:
- Transfer control between threads by waiting "forever" on events or semaphores. To avoid literally waiting forever, use a long timeout (like 60 seconds or longer) and assert that wait() or acquire() returns True. 
- Instead of measuring time, you could do a read/modify/write operation (like incrementing a count variable) and verify that it has the correct result. Unfortunately this may still require sleeping in the middle of the read/modify/write in order to reliably fail.

Reply:
>You call commit in two different threads on the same session. But you don't do it at the same time, do you?

Not at the same time, no. That's not necessary to reproduce the issue.
threading.Thread.join() actually blocks the calling thread until the target thread completes (or a specified timeout occurs). So the first thread completes before the second thread runs.

It doesn't seem unreasonable to also test lock and unlock directly, though I think these tests still have value for their use of the ivi wrapper. I don't understand what value is added by getting an attribute in your proposed test. [The LockSession call blocks until all other threads release their locks](https://www.ni.com/docs/en-US/bundle/ni-scope-c-api-ref/page/scopefunc/cviniscope_locksession.html).

I don't like the idea of checking timing; it may introduce flakiness. Also, I don't think it's necessary to test the locking behavior so rigidly; I don't believe there's anything we could do to the _LockSession call that would allow it to complete without a lock, so long as there isn't a driver bug and our goal isn't to test the driver. As long as we trust that we're calling the right entrypoint, I think it's sufficent to have:

* a function that calls lock(), then unlock()
* 2 threads that call that function
* start thread 1
* start thread 2
* join thread 1
* join thread 2

I'm not opposed to sleeping between the lock and unlock calls.

Reply:
>Also, I don't think it's necessary to test the locking behavior so rigidly

I'm relenting on this point, because the gRPC Interpreter doesn't use the driver functions for locking.
@bkeryan's suggestion of Events was helpful. I've renamed my initial new test and added another test based on this discussion that utilizes an event to keep thread 1 from unlocking before we check the status of thread 2.

Reply:
The test is specific to a bug and as written reproduced the bug, there was a deadlock. Now you confirm it's fixed and this is great.

> It doesn't seem unreasonable to also test lock and unlock directly

That is my point.

> though I think these tests still have value for their use of the ivi wrapper.

Please clarify this statement.

> I don't like the idea of checking timing; it may introduce flakiness.

Fair point. There's other, better ways to test this.

> so long as there isn't a driver bug and our goal isn't to test the driver. 

This is true. There was a time though in which lock/unlock from the Python APIs didn't actually forward it to the driver.

> As long as we trust that we're calling the right entrypoint

`nifake` unit tests can confirm this.

Reply:
> I don't understand what value is added by getting an attribute in your proposed test.

Getting an attribute or making any other call is a way to prove that those functions are doing locking correctly. I proposed getting an attribute because:

1. it's identical across drivers
2. it's usually lightweight

Getting a string type attribute would be best, since it _better_ be grabbing the lock because it needs to do the IVI dance.